src/usr/share/gpgmailer/gpgmailer-daemon.py

```diff
@@ -393,52 +523,87 @@ def sig_term_handler(signal, stack_frame):
     sys.exit(0)
 
 
-config, log_file_handle = build_config_dict()
+def setup_daemon_context(log_file_handle, program_uid, program_gid):
+    """Creates the daemon context.  Specifies daemon permissions, PID file information, and
+    signal handler.
+
+    log_file_handle: The file handle to the log file.
+    program_uid: The system user ID that should own the daemon process.
+    program_gid: The system group ID that should be assigned to the daemon process.
+    Returns the daemon context.
+    """
+    daemon_context = daemon.DaemonContext(
+        working_directory='/',
+        pidfile=pidlockfile.PIDLockFile(
+            os.path.join(SYSTEM_PID_DIR, PROGRAM_PID_DIRS, PID_FILE)),
+        umask=PROGRAM_UMASK,
+        )
+
+    daemon_context.signal_map = {
+        signal.SIGTERM: sig_term_handler,
+        }
+
+    daemon_context.files_preserve = [log_file_handle]
+
+    # Set the UID and GID to 'gpgmailer' user and group.
+    daemon_context.uid = program_uid
+    daemon_context.gid = program_gid
+
+    return daemon_context
+
+
+os.umask(PROGRAM_UMASK)
+program_uid, program_gid = get_user_and_group_ids()
+config, config_helper, logger = read_configuration_and_create_logger(
+    program_uid, program_gid)
+
 try:
+    verify_safe_file_permissions(config, program_uid)
+
     parse_key_config(config)
 
-    create_watch_directories(config)
+    # Re-establish root permissions to create required directories.
+    os.seteuid(os.getuid())
+    os.setegid(os.getgid())
+
+    # Non-root users cannot create files in /run, so create a directory that can be written
+    #   to. Full access to user only.  drwx------ gpgmailer gpgmailer
+    create_directory(SYSTEM_PID_DIR, PROGRAM_PID_DIRS, program_uid, program_gid,
+                     stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
+
+    # Do this relatively last because gpgmailmessage assumes the daemon has started if these
+    #   directories exist.
+    create_spool_directories(program_uid, program_gid)
+
+    # Configuration has been read and directories setup. Now drop permissions forever.
+    drop_permissions_forever(program_uid, program_gid)
 
     # Make sure the sender key isn't going to expire during the first loop iteration.
     expiration_date = time.time() + config['main_loop_duration']
 
     gpg_keyring = gpgkeyring.GpgKeyRing(config['gpg_dir'])
     check_sender_key(gpg_keyring, config, expiration_date)
     check_all_recipient_keys(gpg_keyring, config)
+    verify_signing_config(config)
+
```

Comment:
this isn't actually part of the current code review, but the comment on line 589 has a typo in 'configuration.'